native/src/main/java/io/ballerina/stdlib/io/nativeimpl/CsvChannelUtils.java

```diff
@@ -0,0 +1,171 @@
+
+        BObject characterChannel = ValueCreator.createObjectValue(getIOPackage(), "ReadableByteChannel");
+        CharacterChannelUtils.initCharacterChannel(characterChannel, byteChannel, StringUtils.fromString("UTF-8"));
+
+        BString fs = StringUtils.fromString(",");
+        BString rs = StringUtils.fromString("");
+        BString format = StringUtils.fromString("CSV");
+        BObject textRecordChannel = ValueCreator.createObjectValue(getIOPackage(), "ReadableByteChannel");
+        RecordChannelUtils.initRecordChannel(textRecordChannel, characterChannel, fs, rs, format);
+
+        textRecordChannel.addNativeData(CSV_RETURN_TYPE, typeDesc);
+
+        while (hasNext(textRecordChannel)) {
+
+            return getAllRecords(textRecordChannel, skipHeaders, typeDesc);
+        }
+        return null;
+    }
+
+    public static BStream createCsvAsStream(BString path, BTypedesc typeDesc) {
+        Type describingType = typeDesc.getDescribingType();
+        BObject byteChannel = (BObject) ByteChannelUtils.openReadableFile(path);
+
+        BObject characterChannel = ValueCreator.createObjectValue(getIOPackage(), "ReadableByteChannel");
+        CharacterChannelUtils.initCharacterChannel(characterChannel, byteChannel, StringUtils.fromString("UTF-8"));
+
+        BString fs = StringUtils.fromString(",");
+        BString rs = StringUtils.fromString("");
+        BString format = StringUtils.fromString("CSV");
+        BObject textRecordChannel = ValueCreator.createObjectValue(getIOPackage(), "ReadableByteChannel");
+        RecordChannelUtils.initRecordChannel(textRecordChannel, characterChannel, fs, rs, format);
+        BObject recordIterator = ValueCreator.createObjectValue(getIOPackage(), "CsvIterator");
+
+        recordIterator.addNativeData(CSV_RETURN_TYPE, typeDesc);
+        recordIterator.addNativeData("ITERATOR_NAME", textRecordChannel);
+
+        BStream out = ValueCreator.createStreamValue(
+                TypeCreator.createStreamType(describingType), recordIterator);
+
+        return out;
+    }
+
+    public static Map<String, Object> getStruct(String[] fields, final StructureType structType) {
+        Map<String, Field> internalStructFields = structType.getFields();
+        int fieldLength = internalStructFields.size();
+        Map<String, Object> struct = null;
+        if (fields.length > 0) {
+            Iterator<Map.Entry<String, Field>> itr = internalStructFields.entrySet().iterator();
+            struct = new HashMap<>();
+            for (int i = 0; i < fieldLength; i++) {
+                final Field internalStructField = itr.next().getValue();
+                final int type = internalStructField.getFieldType().getTag();
+                String fieldName = internalStructField.getFieldName();
+                if (fields.length > i) {
+                    String value = fields[i];
+                    switch (type) {
+                        case TypeTags.INT_TAG:
```

Comment:
For non union types, it will be easier and clean to populate the struct in the first switch case, but when the Union types are involved, I think the `getStruct` method might get unnecessarily long and complex.

Reply:
IMO, the logic is not correct here. Following should be the process,
1. Check for union types initially. if not union go to step 3.
2. Validate supported types. If supported type. Get the relevant type description, and pass to step 3.
3. Process the value based on the type description of the base types


Union type is a special case we should handle separately.